src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,141 @@
+            configurable: true);
+
+        private readonly ICodingConventionsManager _conventionsManager = CodingConventionsManagerFactory.CreateCodingConventionsManager();
+
+        public MisplacedUsingsDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.MoveMisplacedUsingsDiagnosticId, s_localizableTitle)
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.CompilationUnit);
+            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.NamespaceDeclaration);
+        }
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            var statement = context.Node;
+            var cancellationToken = context.CancellationToken;
+
+            var option = GetPreferredPlacementOption(context);
+            if (option.Value == UsingPlacementPreference.NoPreference)
+            {
+                return;
+            }
+
+            if (statement is CompilationUnitSyntax compilationUnit
+                && option.Value == UsingPlacementPreference.InsideNamespace
+                && !ShouldSurpressDiagnostic(compilationUnit))
+            {
+                ReportDiagnostics(context, _insideDescriptor, compilationUnit.Usings, option);
+            }
+            else if (statement is NamespaceDeclarationSyntax namespaceDeclaration
+                && option.Value == UsingPlacementPreference.OutsideNamespace)
+            {
+                ReportDiagnostics(context, _outsideDescriptor, namespaceDeclaration.Usings, option);
+            }
+        }
+
+        private CodeStyleOption<UsingPlacementPreference> GetPreferredPlacementOption(SyntaxNodeAnalysisContext context)
+        {
+            var statement = context.Node;
+            var cancellationToken = context.CancellationToken;
+
+            // While running in VisualStudio this should always return an OptionSet.
+            var optionSet = context.Options.GetDocumentOptionSetAsync(statement.SyntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet != null)
+            {
+                return optionSet.GetOption(CSharpCodeStyleOptions.PreferredUsingPlacement);
+            }
+
+            // This code path is for tests only. It relies on the test creating a .editorconfig in the current working directory.
+            var filePath = context.SemanticModel.SyntaxTree.FilePath;
+            var fullPath = string.IsNullOrEmpty(Path.GetDirectoryName(filePath))
+                ? Path.Combine(Environment.CurrentDirectory, filePath)
+                : filePath;
+
+            // Find the correct key name from the .editorconfig storage location
+            var storageLocation = CSharpCodeStyleOptions.PreferredUsingPlacement.StorageLocations
+                .OfType<EditorConfigStorageLocation<CodeStyleOption<UsingPlacementPreference>>>()
+                .FirstOrDefault();
```

Comment:
what is going on here?  In general, there's no way for you to know the implementation of StorageLocations.  This is taking a hard dependency on there being such a config location value here.

Reply:
I was trying to avoid hardcoding the .editorconfig option key into the analyzer. Perhaps a better long term solution is pulling all the .editorconfig option keys into a file we could reference as constants.

Reply:
I guess my point is this:  we already ahve the Options object.  So why are we peaking inside hte options object assumign it has a certain shape.  The Options object is already supposed to be the thing that gives you the values back.  So why is it not just doing all this work inside of itself?